File: Artificial_intelligence/Bot_saves_princess/BotSavesPrincess.py
#input
n = int(input())
matrix = input()

for i in range(0,n-2): #skip useles rows
    input()

matrix = matrix + input()
combinations = ["",""]

#detect princess position and set the proper "zig-zag" combinations
if matrix[0] == 'p':
    combinations[0] = "UP"
    combinations[1] = "LEFT"
else:
    if matrix[n-1] == 'p':
        combinations[0] = "UP"
        combinations[1] = "RIGHT"
    else:
        if matrix[n] == 'p':
            combinations[0] = "DOWN"
            combinations[1] = "LEFT"
        else:
            combinations[0] = "DOWN"
            combinations[1] = "RIGHT"

#move towards the princess
for i in range(0,int((n-1)/2)):
    print (combinations[0])
    print (combinations[1])

File: Artificial_intelligence/Bot_saves_princess/BotSavesPrincess.java
import java.util.Scanner;
/**
 * BotSavesPrincess
 */
public class BotSavesPrincess {

    public static void main(String[] args) {
        Scanner scanner = new Scanner (System.in);
        //input
        int n = scanner.nextInt();
        scanner.nextLine();
        //actualy you need to read only the first and last row (4 corners)
        String input = scanner.nextLine();
        for(int i=1; i<n-1; i++){scanner.nextLine();}   //skip useless lines
        input = input + scanner.nextLine(); 

        String [] combinations = new String [2];
        //detect princess position and set the proper "zig-zag" combinations
        if (input.charAt(0) == 'p'){
            combinations[0] = "UP";
            combinations[1] = "LEFT";
        }
        else
        {
            if (input.charAt(n-1) == 'p'){
                combinations[0] = "UP";
                combinations[1] = "RIGHT";
            }
            else
            {
                if (input.charAt(n) == 'p'){
                    combinations[0] = "DOWN";
                    combinations[1] = "LEFT";
                }
                else
                {
                    combinations[0] = "DOWN";
                    combinations[1] = "RIGHT";
                }
            }
        }
        //move towards the princess
        for(int i=0;i<(n-1)/2;i++){
            System.out.println(combinations[0]);
            System.out.println(combinations[1]);
        }

        
        scanner.close();
    }
}